core/tasks.rs

```
@@ -0,0 +1,273 @@
+  /// ops, from outside of a [`v8::HandleScope`] in a plain `async`` task, or even from within
+  /// another, previously-spawned task.
+  ///
+  /// The task is handed off to be run the next time the event loop is polled, and there are
+  /// no guarantees as to when this may happen.
+  ///
+  /// # Important Notes
+  ///
+  /// The task shares the same [`v8::HandleScope`] as the core event loop, which means that it
+  /// must maintain the scope in a valid state to avoid corrupting or destroying the runtime.
+  ///
+  /// For example, if the code called by this task can raise an exception, the task must ensure
+  /// that it calls that code within a new [`v8::TryCatch`] to avoid the exception leaking to the
+  /// event loop's [`v8::HandleScope`].
+  pub fn spawn<F>(&self, f: F)
+  where
+    F: FnOnce(&mut v8::HandleScope) + 'static,
+  {
+    // SAFETY: we are transmuting Send into a !Send handle but we can guarantee this object will never
+    // leave the current thread because `V8TaskSpawner` is !Send.
+    let task: Box<dyn FnOnce(&mut v8::HandleScope<'_>)> = Box::new(f);
+    let task: Box<dyn FnOnce(&mut v8::HandleScope<'_>) + Send> =
+      unsafe { std::mem::transmute(task) };
+    self.tasks.spawn(task)
+  }
+}
+
+/// Allows for submission of v8 tasks on any thread.
+#[derive(Clone)]
+pub struct V8CrossThreadTaskSpawner {
+  tasks: Arc<V8TaskSpawnerFactory>,
+}
+
+// SAFETY: the underlying V8TaskSpawnerFactory is not Send, but we always submit Send tasks
+// to it from this spawner.
+unsafe impl Send for V8CrossThreadTaskSpawner {}
+
+impl V8CrossThreadTaskSpawner {
+  /// Spawn a task that runs within the [`crate::JsRuntime`] event loop, potentially (but not
+  /// required to be) from a different thread than the runtime is running on.
+  ///
+  /// The task is handed off to be run the next time the event loop is polled, and there are
+  /// no guarantees as to when this may happen.
+  ///
+  /// # Important Notes
```

```suggestion
  /// # Safety
```

It is potentially unsafe if a V8 exception is not handled by the user.